### PR TITLE
Improve ML docs/tests

### LINKED ISSUE #13919: Improve Test Coverage and Documentation for Machine Learning Algorithms
### Feature Description

The `machine_learning/` directory contains several ML algorithm implementations (K-means, Linear Regression, Decision Trees, etc.), but many of these files have limited test coverage, incomplete doctests, and could benefit from more comprehensive documentation. This makes it harder for learners to understand and verify the correctness of implementations.

### Current Issues

1. **Insufficient doctests**: Many ML algorithms lack comprehensive doctests covering edge cases
2. **Missing complexity analysis**: Time and space complexity not documented for most algorithms
3. **Limited examples**: Few practical usage examples with real-world datasets
4. **Incomplete type hints**: Some functions missing proper type annotations
5. **Unclear parameter explanations**: Algorithm parameters not well-documented

### Proposed Enhancements

I propose a systematic improvement of the `machine_learning/` directory:

#### 1. Enhanced Doctests
- Add doctests for all public functions
- Cover edge cases (empty inputs, single data point, etc.)
- Test both typical and boundary conditions
- Include negative test cases (invalid inputs)

#### 2. Comprehensive Documentation
- Add algorithm descriptions with mathematical formulas where appropriate
- Document time and space complexity
- Explain hyperparameters and their effects
- Add references to papers/resources

#### 3. Practical Examples
- Include small example datasets in doctests
- Show typical use cases
- Demonstrate convergence behavior
- Compare with expected outputs

#### 4. Code Quality Improvements
- Complete type hints for all parameters and returns
- Add input validation with proper error messages
- Ensure consistent code style across all ML files
- Add docstring parameters and return value documentation

### Example Files to Improve

- `k_means_clust.py` - Add convergence tests, visualization examples
- `linear_regression.py` - Add tests with known datasets, R² score validation  
- `decision_tree.py` - Test with various tree depths, feature importance
- `gradient_descent.py` - Test convergence with different learning rates
- `naive_bayes.py` - Add probability calculation tests

### Benefits

- **Better learning experience**: Learners can understand algorithms through examples
- **Increased confidence**: Comprehensive tests verify correctness
- **Easier debugging**: Better documentation helps identify issues
- **Professional quality**: Brings ML code to same standard as other directories
- **Reproducibility**: Clear examples make results reproducible

### Suggested Approach

1. Start with most commonly used algorithms (K-means, Linear Regression)
2. Create a template/standard for ML algorithm documentation
3. Systematically apply to all files in `machine_learning/` directory
4. Add GitHub Actions tests to ensure doctest coverage

### Example Enhancement

**Before:**
```python
def k_means(data, k):
    # Basic implementation
    pass
```

**After:**
```python
def k_means(data: np.ndarray, k: int, max_iterations: int = 100) -> tuple[np.ndarray, np.ndarray]:
    """
    K-Means clustering algorithm.
    
    Time Complexity: O(n * k * i * d) where n=samples, k=clusters, i=iterations, d=dimensions
    Space Complexity: O(n * d + k * d)
    
    >>> data = np.array([[1, 2], [1.5, 1.8], [5, 8], [8, 8]])
    >>> centroids, labels = k_means(data, k=2)
    >>> len(centroids)
    2
    """
```

I'm happy to work on this systematic improvement following the repository's contribution guidelines.

### PR DESCRIPTION
**Pull request by MichaeINeumann**
### Describe your change:

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [x] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

## Summary
- improve k-means docstrings/type hints and add convergence doctests
- refactor gradient descent helpers and add targeted unit tests
- add typed linear regression docs/examples and new ML pytest module

## Testing
- .venv/bin/pre-commit run --files machine_learning/k_means_clust.py machine_learning/linear_regression.py machine_learning/gradient_descent.py machine_learning/test_machine_learning_algorithms.py
- .venv/bin/python -m pytest machine_learning/test_machine_learning_algorithms.py

Fixes #13919

### CODE REVIEW

**Review by mindaugl — APPROVED** [review]

**Comment by cclauss** [comment]
@copilot resolve the merge conflicts in this pull request
